src/index/addrindex.cpp

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2017-2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <hash.h>
+#include <index/addrindex.h>
+#include <index/txindex.h>
+#include <shutdown.h>
+#include <primitives/transaction.h>
+#include <script/standard.h>
+#include <ui_interface.h>
+#include <util.h>
+#include <validation.h>
+#include <vector>
+#include <uint256.h>
+
+#include <boost/thread.hpp>
+
+constexpr char DB_ADDRINDEX = 'a';
+std::unique_ptr<AddrIndex> g_addrindex;
+
+/**
+ * Access to the addrindex database (indexes/addrindex/)
+ *
+ * The database stores a block locator of the chain the database is synced to
+ * so that the AddrIndex can efficiently determine the point it last stopped at.
+ * A locator is used instead of a simple hash of the chain tip because blocks
+ * and block index entries may not be flushed to disk until after this database
+ * is updated.
+ */
+class AddrIndex::DB : public BaseIndex::DB
+{
+public:
+    explicit DB(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    // Find all entries in the index for addr_id.
+    // If filter_by_value is true, only returns keys with values equal to value_wanted.
+    bool ReadAddrIndex(const uint64_t addr_id,
+                       std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_found,
+                       const bool filter_by_value = false,
+                       const uint64_t value_wanted = 0);
+
+    bool WriteToIndex(const std::vector<std::pair<uint64_t, CDiskTxPos>> &positions, const uint256 block_hash);
+
+    void RemoveKeys(const std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_to_remove);
+};
+
+AddrIndex::DB::DB(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(GetDataDir() / "indexes" / "addrindex", n_cache_size, f_memory, f_wipe)
+{}
+
+BaseIndex::DB& AddrIndex::GetDB() const { return *m_db; }
+
+bool AddrIndex::DB::ReadAddrIndex(const uint64_t addr_id,
+                                  std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_found,
+                                  const bool filter_by_value,
```

Comment:
Seems that `filter_by_value` functionality isn't used? Also, I'd rename to something like `filter_by_block` because it's not clear what value means and is tightly coupled to the database layout chosen.